codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -464,4 +444,125 @@ public Http2Settings pollSentSettings() {
     public Configuration configuration() {
         return frameWriter.configuration();
     }
+
+    /**
+     * Wrap a DATA frame so it can be written subject to flow-control. Note that this implementation assumes it
+     * only writes padding once for the entire payload as opposed to writing it once per-frame. This makes the
+     * {@link #size} calculation deterministic thereby greatly simplifying the implementation.
+     * <p>
+     * If frame-splitting is required to fit within max-frame-size and flow-control constraints we ensure that
+     * the passed promise is not completed until last frame write.
+     * </p>
+     */
+    private final class FlowControlledData extends FlowControlledPayloadBase {
+        private final ByteBuf data;
+        private int size;
+
+        private FlowControlledData(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding,
+                                    boolean endOfStream, ChannelPromise promise) {
+            super(ctx, padding, endOfStream, streamId, promise);
+            this.data = data;
+            size = data.readableBytes() + padding;
```

Comment:
I originally tried this but it broke because DefaultFrameWriter by copying the buffer doesn't actually cause the reader index to be updated. This could and maybe should be fixed but it made me think a bit about the contract between the two. If frame writer weren't copying then it would simply write(data) without a flush which would also mean its reader index wouldn't update until after the flush which can occur much later. Given that we need to check size() synchronously in the flow-controller this becomes problematic